pandas/core/groupby/grouper.py

```diff
@@ -938,7 +932,7 @@ def _convert_grouper(axis: Index, grouper):
             return grouper.reindex(axis)._values
     elif isinstance(grouper, MultiIndex):
         return grouper._values
-    elif isinstance(grouper, (list, tuple, Series, Index, np.ndarray)):
+    elif isinstance(grouper, (list, tuple, Index, Categorical, np.ndarray)):
```

Comment:
why is this changed?

Reply:
First, `Series` has no effect, because a prior `elif` [clause](https://github.com/pandas-dev/pandas/blob/8d10ab4facf7225862150ca7695f60f8f5402bfe/pandas/core/groupby/grouper.py#L934) catches instances of `Series`.  The [first commit](https://github.com/pandas-dev/pandas/pull/44180/commits/8d10ab4facf7225862150ca7695f60f8f5402bfe) of this PR removes `Series` to clean up the line before other changes.

Second, inserting `Categorical` replaces and tightens the check formerly performed by the following [line](https://github.com/pandas-dev/pandas/blob/8d10ab4facf7225862150ca7695f60f8f5402bfe/pandas/core/groupby/grouper.py#L890) in `get_grouper()`:

```python
if is_categorical_dtype(gpr) and len(gpr) != obj.shape[axis]:
```

Removing that `if` block and inserting `Categorical` into the `isinstance` check is the crux of the bug fix.  By removing the `if` block, we in particular allow `Series` groupers with categorical data type and length unequal to that of the grouping axis to proceed and eventually be reindexed.  By inserting `Categorical`, we continue to require that groupers of that type (which are array-like and lack an index) have the same length as the axis of grouping.